torch/export/export_schema.py

```diff
@@ -0,0 +1,319 @@
+#       These are faithful capture of tensor's detail in pytorch's executions during tracing
+#       However, it's up to downstream system on how to utilized these fields
+#       In another word, these feilds are suggestive, rather than mandatory.
+
+@dataclass
+class TensorMeta:
+    dtype: ScalarType
+    sizes: List[SymInt]
+
+    # needed for training
+    requires_grad: bool
+
+    # !!! see description above, there are subtle difference on how these fields should be interpreted
+    device: Device
+    strides: List[SymInt]
+    storage_offset: int
+    layout: Layout
+
+    # !!! memory_format is not tensor property, but it can only used as operator argument
+    # memory_format: MemoryFormat
+
+
+@dataclass
+class Buffer:
+    # !!! TODO: need to define endianness for the buffer
+    buffer: bytes
+
+
+@dataclass
+class ExternalBuffer:
+    # !!! Example from onnx
+    #   // Data can be stored inside the protobuf file using type-specific fields or raw_data.
+    #   // Alternatively, raw bytes data can be stored in an external file, using the external_data field.
+    #   // external_data stores key-value pairs describing data location. Recognized keys are:
+    #   // - "location" (required) - POSIX filesystem path relative to the directory where the ONNX
+    #   //                           protobuf model was stored
+    #   // - "offset" (optional) - position of byte at which stored data begins. Integer stored as string.
+    #   //                         Offset values SHOULD be multiples 4096 (page size) to enable mmap support.
+    #   // - "length" (optional) - number of bytes containing data. Integer stored as string.
+    #   // - "checksum" (optional) - SHA1 digest of file specified in under 'location' key.
+    #   repeated StringStringEntryProto external_data
+    location: str
+    offset: str     # !!! Consider using int, but int has int_max limitation
+    length: str     # !!! Consider using int, but int has int_max limitation
+    checksum: str
+
+
+@dataclass
+class Storage:
+    class DataLocation(Enum):
+        Internal = auto()
+        External = auto()
+
+    data_location: DataLocation
+    data: Union[Buffer, ExternalBuffer]
+
+
+# This is a concrete tensor backed by storage
+@dataclass
+class Tensor:
+    # storage
+    storage: Storage
+
+    # metadata
+    meta: TensorMeta
+
+
+################################################################################
+# Following section is the defining the schema of 3 level construct: GraphModule, Graph, Node
+
+# IValue has no corresponding class in fx
+# IValue is the "values" that are passed between nodes in the graph
+# !!! Assumption: Only Tensor can be passed between nodes, and not other int/float/bool types...
+# IValue is a named virtual tensor, with an optional TensorMeta that describes the properties of the tensor
+# !!! Consider using a more descriptive name, e.g. TensorValue, TensorPlaceholder, TensorArgument, etc.
+@dataclass
+class IValue:
```

Comment:
Answering @zhxchen17 's questions on why we need a construct for IValue. 

The major question is that there is no corresponding construct of IValue in fx, so why do we need to invent a new construct for IValue? 

This is actually an important design decision and has a lot of implications, so let me reason through both options: with-IValue vs. without-IValue. 

Basically, IValue is an abstraction of intermediate values that flows across the Graph, aka the edges in the graph. Notice that FX graph doesn't have a concept of "Edge", but actually, fx graph has an implicit assumption that all the edges is just a python object. As a short hand expression, fx uses producer node's name to represent the python object itself. This is a very smart way to hide the existence of edge, and it works fine in python. 

Due to this design choice, FX's graph has a few consequences:
Pros:
- Simplicity. No need for the abstraction of Edge! 
- Flexible. Any python object can be passed between nodes.
Cons:
- To represent multiple tensor returns, it must be packed into a list or a tuple, and unpacked by a subsequent getitem node. 
- Nodes that returns multiple outputs acctually doesn't have FakeTensor node.meta['val'] populated. It needs to be found in the subsequent getitem node. 


With the introduction of IValue, we have a few consequences:
Pros:
- Edge has an explict abstration. This allows us to apply constrains on what can be passed between nodes. This is important for downstream C++ system, as it cannot handle arbitory unknown types. ATM, we want to make a constrain that only Tensor and SymInt are allowed to be passed between nodes. 
- For the handling of multile output, the graph doesn't need the additional getitem node to unpack. Multiple outputs can be represented as multiple IValues. In fact, every intemediate tensor has a corresponding IValue instance. 
- FakeTensor can be directly stored together with to IValue. This is more natural than storing FakeTensor in a node.meta. After all, FakeTensor is descriptor on Tensor, not on Node. 
Cons:
- This introduces inconsistency between exported graph format and FX graph format, which adds cognative burden to the developers. (but this is a very common pattern used in TorchScript and other compiler frameworks)

Reply:
cc @qihqi

Reply:
> ATM, we want to make a constrain that only Tensor and SymInt are allowed to be passed between nodes.

So things like int, float, bool parameters of ops generally considered constants and they cannot come as primary inputs of the graph? For example, if I have upsample op where output size is one of the inputs of the model, then such input must be captured as symint?